pkg/validation-webhook/groupsnapshot.go

```diff
@@ -0,0 +1,244 @@
+			reviewResponse.Allowed = false
+			reviewResponse.Result.Message = err.Error()
+			return reviewResponse
+		}
+	}
+	// Enforce strict validation for all CREATE requests. Immutable checks don't apply for CREATE requests.
+	// Enforce strict validation for UPDATE requests where old is valid and passes immutability check.
+	if err := ValidateV1Alpha1GroupSnapshotContent(groupSnapcontent); err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+	}
+	return reviewResponse
+}
+
+func decideGroupSnapshotClassV1Alpha1(snapClass, oldSnapClass *volumegroupsnapshotv1alpha1.VolumeGroupSnapshotClass, lister groupSnapshotListers.VolumeGroupSnapshotClassLister) *v1.AdmissionResponse {
+	reviewResponse := &v1.AdmissionResponse{
+		Allowed: true,
+		Result:  &metav1.Status{},
+	}
+
+	// Only Validate when a new group snapshot class is being set as a default.
+	if snapClass.Annotations[utils.IsDefaultSnapshotClassAnnotation] != "true" {
+		return reviewResponse
+	}
+
+	// If Old group snapshot class has this, then we can assume that it was validated if driver is the same.
+	if oldSnapClass.Annotations[utils.IsDefaultSnapshotClassAnnotation] == "true" && oldSnapClass.Driver == snapClass.Driver {
+		return reviewResponse
+	}
+
+	ret, err := lister.List(labels.Everything())
+	if err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+		return reviewResponse
+	}
+
+	for _, snapshotClass := range ret {
+		if snapshotClass.Annotations[utils.IsDefaultSnapshotClassAnnotation] != "true" {
+			continue
+		}
+		if snapshotClass.Driver == snapClass.Driver {
+			reviewResponse.Allowed = false
+			reviewResponse.Result.Message = fmt.Sprintf("default group snapshot class: %v already exists for driver: %v", snapshotClass.Name, snapClass.Driver)
+			return reviewResponse
+		}
+	}
+
+	return reviewResponse
+}
+
+func checkGroupSnapshotImmutableFieldsV1Alpha1(groupSnapshot, oldGroupSnapshot *volumegroupsnapshotv1alpha1.VolumeGroupSnapshot) error {
+	if groupSnapshot == nil {
+		return fmt.Errorf("VolumeGroupSnapshot is nil")
+	}
+	if oldGroupSnapshot == nil {
+		return fmt.Errorf("old VolumeGroupSnapshot is nil")
+	}
+
+	source := groupSnapshot.Spec.Source
+	oldSource := oldGroupSnapshot.Spec.Source
+
+	if !reflect.DeepEqual(source.Selector, oldSource.Selector) {
+		return fmt.Errorf("Spec.Source.Selector is immutable but was changed from %s to %s", oldSource.Selector, source.Selector)
+	}
+	if !reflect.DeepEqual(source.VolumeGroupSnapshotContentName, oldSource.VolumeGroupSnapshotContentName) {
+		return fmt.Errorf("Spec.Source.VolumeGroupSnapshotContentName is immutable but was changed from %s to %s", strPtrDereference(oldSource.VolumeGroupSnapshotContentName), strPtrDereference(source.VolumeGroupSnapshotContentName))
+	}
+
+	return nil
+}
+
+func checkGroupSnapshotContentImmutableFieldsV1Alpha1(groupSnapcontent, oldGroupSnapcontent *volumegroupsnapshotv1alpha1.VolumeGroupSnapshotContent) error {
```

Comment:
I think this function also needs to deal with `VolumeGroupSnapshotRef`:

```
type VolumeGroupSnapshotContentSpec struct {
	// VolumeGroupSnapshotRef specifies the VolumeGroupSnapshot object to which this
	// VolumeGroupSnapshotContent object is bound.
	// VolumeGroupSnapshot.Spec.VolumeGroupSnapshotContentName field must reference to
	// this VolumeGroupSnapshotContent's name for the bidirectional binding to be valid.
	// For a pre-existing VolumeGroupSnapshotContent object, name and namespace of the
	// VolumeGroupSnapshot object MUST be provided for binding to happen.
	// This field is immutable after creation.
	// Required.
	VolumeGroupSnapshotRef core_v1.ObjectReference `json:"volumeGroupSnapshotRef" protobuf:"bytes,1,opt,name=volumeGroupSnapshotRef"`
...
```

Reply:
Just note that in the case of pre-provisioned group snapshot, the "uid" field of VolumeGroupSnapshotRef should be added by the snapshot controller later while the "name" and "namespace" were created by the admin.  So to ensure immutability, we need to check the individual fields within VolumeGroupSnapshotRef.